crates/services/txpool_v2/src/heavy_async_processing.rs

```diff
@@ -0,0 +1,195 @@
+use std::{
+    future::Future,
+    sync::Arc,
+};
+use tokio::sync::Semaphore;
+
+pub struct HeavyAsyncProcessor {
+    rayon_thread_pool: rayon::ThreadPool,
+    semaphore: Arc<Semaphore>,
+}
```

Comment:
Related to https://github.com/FuelLabs/fuel-core/pull/2162#discussion_r1783473306 but a bit orthogonal since that comment still assumes we use the dedicated thread pool when consuming messages.

Reply:
Now it's also used by the insertion/verifications of transactions in this PR. The need is that we want to have a way to not block the current executor to do their thing but also to have a limit on the resources we dedicate to specific tasks. In this case we place arbitrary number of threads and queue length on tx verif/insert and some arbitrary on p2p sync. This two operations can take a lot of computational power/waiting a lot that's why this is those ones that are separated.

This file isn't an exact copy as it works with async code which is not the same as in p2p but I agree that it's very similar and I asked where I can place it to share the two and it seems that there is no good place for this.

Reply:
Alright, but are there any blocking calls? That seems like what `tokio::spawn_blocking` is made for. When it comes to limiting the number of active futures I agree that it makes sense to use the `Semaphore`. I am more skeptical about the rayon part.